env/context.go

```diff
@@ -5,59 +5,184 @@
 package env
 
 import (
+	"fmt"
 	"net"
+	"os"
+	"path/filepath"
 	"sync"
 
 	"github.com/keybase/client/go/libkb"
 	"github.com/keybase/go-framed-msgpack-rpc/rpc"
 )
 
-// Context is an implementation for libkbfs.Context
-type Context struct {
+const (
+	kbfsSocketFile = "kbfsd.sock"
+)
+
+// Context defines the environment for this package
+type Context interface {
+	GetRunMode() libkb.RunMode
+	GetLogDir() string
+	GetDataDir() string
+	ConfigureSocketInfo() (err error)
+	GetSocket(clearError bool) (net.Conn, rpc.Transporter, bool, error)
+	NewRPCLogFactory() *libkb.RPCLogFactory
+	// TODO: uncomment once we've added changes to allow gobind to create a
+	// KBFS context from a *libkb.GlobalContext
+	//GetKBFSSocket(clearError bool) (net.Conn, rpc.Transporter, bool, error)
+	//BindToKBFSSocket() (net.Listener, error)
+}
+
+// KBFSContext is an implementation for libkbfs.Context
+type KBFSContext struct {
 	g *libkb.GlobalContext
+	// protects the socket primitives below
+	kbfsSocketMtx     sync.RWMutex
+	kbfsSocket        libkb.Socket
+	kbfsSocketWrapper *libkb.SocketWrapper
 }
 
-var libkbOnce sync.Once
+var _ Context = (*KBFSContext)(nil)
 
-// NewContext constructs a context
-func NewContext() *Context {
-	// TODO: Remove direct use of libkb.G
-	libkbOnce.Do(func() {
-		libkb.G.Init()
-		libkb.G.ConfigureConfig()
-		libkb.G.ConfigureLogging()
-		libkb.G.ConfigureCaches()
-		libkb.G.ConfigureMerkleClient()
+func (c *KBFSContext) initKBFSSocket() {
+	c.kbfsSocketMtx.Lock()
+	defer c.kbfsSocketMtx.Unlock()
+	log := c.g.Log
+	bindFile := c.getKBFSSocketFile()
```

Comment:
Does iOS / bind stuff work through this too, or is it just on desktop? I guess I'm lacking a lot of context.

Reply:
For now, I'm going to avoid actually binding to the socket on mobile, since we don't need to make `kbgit` work on mobile yet. Eventually if we want to support it though, we'll need a loopback listener like the service instantiates for KBFS.

Reply:
I see; thanks for explaining!